File: pkg/render/render.go
// Package render produces the Markdown PR comment body from comparison results.
package render

import (
	"cmp"
	"fmt"
	"slices"
	"strings"

	"github.com/vvoland/gha-pin-diff/pkg/compare"
	"github.com/vvoland/gha-pin-diff/pkg/pinverify"
)

// Marker is the HTML comment used to identify bot comments.
const Marker = "<!-- gha-pin-diff -->"

// MaxCommitsShown is the maximum number of commits displayed per action.
const MaxCommitsShown = 15

// Comment renders the full Markdown comment body for the given comparison results.
// Returns empty string if there are no results and no mismatches.
func Comment(results []compare.Result, mismatches []pinverify.Mismatch) string {
	if len(results) == 0 && len(mismatches) == 0 {
		return ""
	}

	// Deduplicate and partition into pin-only vs changed.
	results = dedup(results)
	var pinOnly, changed []compare.Result
	for _, r := range results {
		if isPinOnly(r) {
			pinOnly = append(pinOnly, r)
		} else {
			changed = append(changed, r)
		}
	}

	var b strings.Builder
	b.WriteString(Marker)
	b.WriteString("\n## 🔄 Action Pin Diff\n")

	if len(mismatches) > 0 {
		renderMismatches(&b, mismatches)
	}

	for _, r := range changed {
		b.WriteByte('\n')
		renderResult(&b, r)
	}

	if len(pinOnly) > 0 {
		renderPinOnly(&b, pinOnly)
	}

	return b.String()
}

func dedup(results []compare.Result) []compare.Result {
	// Sort by action, then refs for grouping, then file for stable output.
	slices.SortFunc(results, func(a, b compare.Result) int {
		if c := cmp.Compare(a.Update.Action, b.Update.Action); c != 0 {
			return c
		}
		if c := cmp.Compare(a.Update.OldRef, b.Update.OldRef); c != 0 {
			return c
		}
		if c := cmp.Compare(a.Update.NewRef, b.Update.NewRef); c != 0 {
			return c
		}
		return cmp.Compare(a.Update.File, b.Update.File)
	})

	return slices.CompactFunc(results, func(a, b compare.Result) bool {
		return a.Update.Action == b.Update.Action &&
			a.Update.OldRef == b.Update.OldRef &&
			a.Update.NewRef == b.Update.NewRef
	})
}

// isPinOnly reports whether a result is a digest pin with no version change.
func isPinOnly(r compare.Result) bool {
	return r.Err == nil && r.TotalCommits == 0
}

func renderPinOnly(b *strings.Builder, results []compare.Result) {
	b.WriteString("\n### 📌 Pinned (digest unchanged)\n")
	b.WriteString("\n| Action | Version | Digest |\n")
	b.WriteString("|--------|---------|--------|\n")
	for _, r := range results {
		u := r.Update
		tag := u.NewTag
		if tag == "" {
			tag = u.OldTag
		}
		if tag == "" {
			tag = shortRef(u.NewRef)
		}
		commitURL := fmt.Sprintf("https://github.com/%s/commit/%s", actionOwnerRepo(u.Action), u.NewRef)
		fmt.Fprintf(b, "| [`%s`](%s) | `%s` | [`%s`](%s) |\n", u.Action, actionRepoURL(u.Action), tag, shortRef(u.NewRef), commitURL)
	}
}

func renderMismatches(b *strings.Builder, mismatches []pinverify.Mismatch) {
	b.WriteString("\n### ⚠️ Tag / SHA Mismatch\n")
	b.WriteString("\nThe following pins reference a SHA that does not match the tag in the comment:\n")
	b.WriteString("\n| Action | Tag | Expected SHA | Pinned SHA |\n")
	b.WriteString("|--------|-----|-------------|------------|\n")
	for _, m := range mismatches {
		u := m.Update
		tagURL := fmt.Sprintf("https://github.com/%s/releases/tag/%s", actionOwnerRepo(u.Action), m.Tag)
		expectURL := fmt.Sprintf("https://github.com/%s/commit/%s", actionOwnerRepo(u.Action), m.ExpectSHA)
		pinnedURL := fmt.Sprintf("https://github.com/%s/commit/%s", actionOwnerRepo(u.Action), u.NewRef)
		fmt.Fprintf(b, "| [`%s`](%s) | [`%s`](%s) | [`%s`](%s) | [`%s`](%s) |\n",
			u.Action, actionRepoURL(u.Action),
			m.Tag, tagURL,
			shortRef(m.ExpectSHA), expectURL,
			shortRef(u.NewRef), pinnedURL)
	}
}

func renderResult(b *strings.Builder, r compare.Result) {
	u := r.Update
	actionRepo := actionRepoURL(u.Action)

	// Header: action name with version range
	fmt.Fprintf(b, "### [`%s`](%s)", u.Action, actionRepo)
	if u.OldTag != "" || u.NewTag != "" {
		old := u.OldTag
		if old == "" {
			old = shortRef(u.OldRef)
		}
		new := u.NewTag
		if new == "" {
			new = shortRef(u.NewRef)
		}
		fmt.Fprintf(b, " `%s` → `%s`", old, new)
	}
	b.WriteByte('\n')

	if r.Err != nil {
		fmt.Fprintf(b, "\n⚠️ Could not fetch comparison: %v\n", r.Err)
		fmt.Fprintf(b, "\n[View diff manually](https://github.com/%s/compare/%s...%s)\n",
			actionOwnerRepo(u.Action), u.OldRef, u.NewRef)
		return
	}

	// Summary line
	commitWord := "commits"
	if r.TotalCommits == 1 {
		commitWord = "commit"
	}
	fmt.Fprintf(b, "\n**%d %s**", r.TotalCommits, commitWord)
	if r.CompareURL != "" {
		fmt.Fprintf(b, " · [Compare](%s)", r.CompareURL)
	}
	b.WriteByte('\n')

	if len(r.Commits) == 0 {
		return
	}

	// Commit table
	ownerRepo := actionOwnerRepo(u.Action)
	b.WriteString("\n| SHA | Message | Date |\n")
	b.WriteString("|-----|---------|------|\n")

	shown := r.Commits
	if len(shown) > MaxCommitsShown {
		shown = shown[len(shown)-MaxCommitsShown:]
	}

	for _, c := range shown {
		sha := shortRef(c.SHA)
		commitURL := fmt.Sprintf("https://github.com/%s/commit/%s", ownerRepo, c.SHA)
		date := ""
		if !c.Date.IsZero() {
			date = c.Date.Format("2006-01-02")
		}
		msg := escapeMarkdown(c.Message)
		if len(msg) > 80 {
			msg = msg[:77] + "..."
		}
		fmt.Fprintf(b, "| [`%s`](%s) | %s | %s |\n", sha, commitURL, msg, date)
	}

	if r.TotalCommits > MaxCommitsShown {
		fmt.Fprintf(b, "\n*Showing %d of %d commits. [View all](%s)*\n",
			MaxCommitsShown, r.TotalCommits, r.CompareURL)
	}
}

func actionRepoURL(action string) string {
	return "https://github.com/" + actionOwnerRepo(action)
}

func actionOwnerRepo(action string) string {
	owner, rest, ok := strings.Cut(action, "/")
	if !ok {
		return action
	}
	repo, _, _ := strings.Cut(rest, "/")
	return owner + "/" + repo
}

func escapeMarkdown(s string) string {
	s = strings.ReplaceAll(s, "|", "\\|")
	s = strings.ReplaceAll(s, "\n", " ")
	return s
}

func shortRef(ref string) string {
	if len(ref) > 7 {
		return ref[:7]
	}
	return ref
}

File: src/render/render.ts
import type { Result } from "../compare/compare.js";
import { updateOwnerRepo } from "../compare/compare.js";
import type { Mismatch } from "../pinverify/verify.js";
import type { DigestMismatch } from "../registry/verify.js";
import type { ActionUpdate } from "../diffparser/parser.js";

/** HTML comment used to identify bot comments. */
export const MARKER = "<!-- gha-pin-diff -->";

/** Maximum number of commits displayed per action. */
export const MAX_COMMITS_SHOWN = 15;

/**
 * Renders the full Markdown comment body for the given comparison results.
 * Returns empty string if there are no results and no mismatches.
 */
export function comment(
  results: Result[],
  mismatches: Mismatch[],
  digestMismatches: DigestMismatch[] = []
): string {
  if (
    results.length === 0 &&
    mismatches.length === 0 &&
    digestMismatches.length === 0
  )
    return "";

  results = dedup(results);
  const pinOnly: Result[] = [];
  const images: Result[] = [];
  const changed: Result[] = [];
  for (const r of results) {
    if (isImage(r)) {
      images.push(r);
      if (updateOwnerRepo(r.update) !== null && (r.err || r.totalCommits > 0)) {
        changed.push(r);
      }
    } else if (isPinOnly(r)) {
      pinOnly.push(r);
    } else {
      changed.push(r);
    }
  }

  let b = "";
  b += MARKER;
  b += "\n## 🔄 Action Pin Diff\n";

  if (mismatches.length > 0) {
    b += renderMismatches(mismatches);
  }

  if (digestMismatches.length > 0) {
    b += renderDigestMismatches(digestMismatches);
  }

  for (const r of changed) {
    b += "\n";
    b += renderResult(r);
  }

  if (images.length > 0) {
    b += renderImages(images);
  }

  if (pinOnly.length > 0) {
    b += renderPinOnly(pinOnly);
  }

  return b;
}

function dedup(results: Result[]): Result[] {
  const sorted = [...results].sort((a, b) => {
    const aKey = updateKey(a.update);
    const bKey = updateKey(b.update);
    if (aKey !== bKey) return aKey < bKey ? -1 : 1;
    return a.update.file < b.update.file
      ? -1
      : a.update.file > b.update.file
        ? 1
        : 0;
  });

  return sorted.filter(
    (r, i) =>
      i === 0 ||
      updateKey(r.update) !== updateKey(sorted[i - 1].update)
  );
}

function updateKey(u: ActionUpdate): string {
  return [
    u.action,
    u.repo ?? "",
    u.oldRef,
    u.newRef,
    u.oldTag,
    u.newTag,
    u.oldDigest ?? "",
    u.newDigest ?? "",
  ].join("\0");
}

function isPinOnly(r: Result): boolean {
  return r.err === null && r.totalCommits === 0;
}

/**
 * Reports whether r is a Docker image update. Compose images always carry a
 * home URL, including GHCR images that also map to a GitHub repository.
 */
function isImage(r: Result): boolean {
  return !!r.update.homeURL;
}

function ownerRepoStr(update: Pick<ActionUpdate, "action" | "repo">): string | null {
  const parsed = updateOwnerRepo(update);
  return parsed ? parsed.join("/") : null;
}

function renderPinOnly(results: Result[]): string {
  let b = "\n### 📌 Pinned (digest unchanged)\n";
  b += "\n| Action | Version | Digest |\n";
  b += "|--------|---------|--------|\n";
  for (const r of results) {
    const u = r.update;
    let tag = u.newTag || u.oldTag;
    if (!tag) tag = shortRef(u.newRef);
    const repo = ownerRepoStr(u);
    const commitURL = repo ? `https://github.com/${repo}/commit/${u.newRef}` : "";
    b += `| ${renderLabel(u)} | \`${tag}\` | ${renderRef(shortRef(u.newRef), commitURL)} |\n`;
  }
  return b;
}

function renderImages(results: Result[]): string {
  let b = "\n### 🐳 Docker Images\n";
  b += "\n| Image | Old | New |\n";
  b += "|-------|-----|-----|\n";
  for (const r of results) {
    const u = r.update;
    const old = imageRef(u.oldTag || shortRef(u.oldRef), u.oldDigest);
    const newRef = imageRef(u.newTag || shortRef(u.newRef), u.newDigest);
    b += `| ${renderLabel(u)} | \`${old}\` | \`${newRef}\` |\n`;
  }
  return b;
}

/**
 * Formats an image reference as `tag@digest`, appending the shortened digest
 * only when the image is digest-pinned. The digest is the immutable pin, so it
 * is the part worth surfacing in review.
 */
function imageRef(tag: string, digest?: string): string {
  return digest ? `${tag}@${shortDigest(digest)}` : tag;
}

/** Shortens a `sha256:<hash>` digest to its algorithm and first 12 hex chars. */
function shortDigest(digest: string): string {
  const colon = digest.indexOf(":");
  if (colon < 0) return shortRef(digest);
  return `${digest.substring(0, colon)}:${digest.substring(colon + 1, colon + 13)}`;
}

function renderMismatches(mismatches: Mismatch[]): string {
  let b = "\n### ⚠️ Tag / SHA Mismatch\n";
  b += "\nThe following pins reference a SHA that does not match the tag in the comment:\n";
  b += "\n| Action | Tag | Expected SHA | Pinned SHA |\n";
  b += "|--------|-----|-------------|------------|\n";
  for (const m of mismatches) {
    const u = m.update;
    const or = ownerRepoStr(u);
    const tagURL = or ? `https://github.com/${or}/releases/tag/${m.tag}` : "";
    const expectURL = or ? `https://github.com/${or}/commit/${m.expectSHA}` : "";
    const pinnedURL = or ? `https://github.com/${or}/commit/${u.newRef}` : "";
    b += `| ${renderLabel(u)} | ${renderRef(m.tag, tagURL)} | ${renderRef(shortRef(m.expectSHA), expectURL)} | ${renderRef(shortRef(u.newRef), pinnedURL)} |\n`;
  }
  return b;
}

function renderDigestMismatches(mismatches: DigestMismatch[]): string {
  let b = "\n### ⚠️ Image Tag / Digest Mismatch\n";
  b += "\nThe following images pin a digest that the registry no longer resolves the tag to:\n";
  b += "\n| Image | Tag | Registry Digest | Pinned Digest |\n";
  b += "|-------|-----|-----------------|---------------|\n";
  for (const m of mismatches) {
    const u = m.update;
    b += `| ${renderLabel(u)} | \`${m.tag}\` | \`${shortDigest(m.expectDigest)}\` | \`${shortDigest(u.newDigest ?? "")}\` |\n`;
  }
  return b;
}

function renderResult(r: Result): string {
  const u = r.update;
  const repo = ownerRepoStr(u);

  let b = `### ${renderLabel(u)}`;
  if (u.oldTag || u.newTag) {
    const old = u.oldTag || shortRef(u.oldRef);
    const newTag = u.newTag || shortRef(u.newRef);
    b += ` \`${old}\` → \`${newTag}\``;
  }
  b += "\n";

  if (r.err) {
    b += `\n⚠️ Could not fetch comparison: ${r.err.message}\n`;
    b += `\nRefs: \`${shortRef(u.oldRef)}\` → \`${shortRef(u.newRef)}\`\n`;
    if (repo) {
      b += `\n[View diff manually](https://github.com/${repo}/compare/${u.oldRef}...${u.newRef})\n`;
    }
    return b;
  }

  const commitWord = r.totalCommits === 1 ? "commit" : "commits";
  b += `\n**${r.totalCommits} ${commitWord}**`;
  if (r.compareURL) {
    b += ` · [Compare](${r.compareURL})`;
  }
  b += "\n";

  if (r.commits.length === 0) return b;

  const or = ownerRepoStr(u);
  b += "\n| SHA | Message | Date |\n";
  b += "|-----|---------|------|\n";

  let shown = r.commits;
  if (shown.length > MAX_COMMITS_SHOWN) {
    shown = shown.slice(shown.length - MAX_COMMITS_SHOWN);
  }

  for (const c of shown) {
    const sha = shortRef(c.sha);
    const commitURL = or ? `https://github.com/${or}/commit/${c.sha}` : "";
    const date = c.date ? formatDate(c.date) : "";
    let msg = escapeMarkdown(c.message);
    if (msg.length > 80) {
      msg = msg.substring(0, 77) + "...";
    }
    b += `| ${renderRef(sha, commitURL)} | ${msg} | ${date} |\n`;
  }

  if (r.totalCommits > MAX_COMMITS_SHOWN) {
    b += `\n*Showing ${MAX_COMMITS_SHOWN} of ${r.totalCommits} commits. [View all](${r.compareURL})*\n`;
  }

  return b;
}

function actionRepoURL(
  update: Pick<ActionUpdate, "action" | "repo" | "homeURL">
): string | null {
  const repo = ownerRepoStr(update);
  if (repo) return "https://github.com/" + repo;
  return update.homeURL ?? null;
}

function renderLabel(
  update: Pick<ActionUpdate, "action" | "repo" | "homeURL">
): string {
  return renderRef(update.action, actionRepoURL(update));
}

function renderRef(text: string, url: string | null): string {
  return url ? `[\`${text}\`](${url})` : `\`${text}\``;
}

function escapeMarkdown(s: string): string {
  return s.replaceAll("|", "\\|").replaceAll("\n", " ");
}

function shortRef(ref: string): string {
  return ref.length > 7 ? ref.substring(0, 7) : ref;
}

function formatDate(d: Date): string {
  const y = d.getUTCFullYear();
  const m = String(d.getUTCMonth() + 1).padStart(2, "0");
  const day = String(d.getUTCDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
}
